File: 0133-clone-graph/0133-clone-graph.cs
/*
// Definition for a Node.
public class Node {
    public int val;
    public IList<Node> neighbors;

    public Node() {
        val = 0;
        neighbors = new List<Node>();
    }

    public Node(int _val) {
        val = _val;
        neighbors = new List<Node>();
    }

    public Node(int _val, List<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}
*/

public class Solution {
    public Node CloneGraph(Node node) {
        if (node == null) return null;

        Dictionary<Node, int> visited = new Dictionary<Node, int>(); // Track visited nodes
        Queue<Node> q = new Queue<Node>(); // BFS queue
        Dictionary<Node, Node> mp = new Dictionary<Node, Node>(); // Mapping original nodes to cloned nodes

        Node cloneGraph = new Node(node.val); // Create first clone
        q.Enqueue(node);
        mp[node] = cloneGraph;

        while (q.Count > 0) {
            Node orig = q.Dequeue();
            Node cpy;

            if (!mp.ContainsKey(orig))
                mp[orig] = new Node(orig.val);
            
            cpy = mp[orig];
            List<Node> nbrs = new List<Node>();
            visited[orig] = 1;

            foreach (var i in orig.neighbors) {
                Node n;
                if (!visited.ContainsKey(i)) q.Enqueue(i);
                if (!mp.ContainsKey(i)) mp[i] = new Node(i.val);
                n = mp[i];
                nbrs.Add(n);
            }
            
            cpy.neighbors = nbrs;
        }
        return cloneGraph;
    }
}

File: 0133-clone-graph/0133-clone-graph.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/

class Solution {
public:
    Node *cloneGraph(Node *node)
    {
        if (node == NULL)
        {
            return node;
        }
        map<Node *, int> visited;
        queue<Node *> q;
        map<Node *, Node *> mp;
        Node *cloneGraph = new Node(node->val);
        q.push(node);
        mp[node] = cloneGraph;

        while (!q.empty())
        {
            auto f = q.front();
            q.pop();
            Node *orig = f;
            Node *cpy;
            if (mp.find(orig) == mp.end())
            {
                mp[orig] = new Node(orig->val);
            }
            cpy = mp[orig];
            vector<Node *> nbrs;
            visited[orig] = 1;

            for (auto i : orig->neighbors)
            {
                Node *n;
                if (visited.find(i) == visited.end()) q.push(i);
                if(mp.find(i) == mp.end()) mp[i] = new Node(i->val);
                n = mp[i];
                nbrs.push_back(n);
            }
            cpy->neighbors = nbrs;
        }
        return cloneGraph;
    }
};